sql: panic: could not find overload for binary expression minus

```
values (true, NULL, true, (7252766201640790092 + current_date()), false, case when case when true then false else false end then -0.5284822756467706 else (case when true then NULL else -0.41697856420581636 end - case when NULL then 1.4034371360919229 else ln(NULL) end) end, 3073126136, case when false then e'{",uChz": {"a": {"J_df": [true, 2.1106438556494282], "`|0a`;R": {}}, "nE$\'$": []}, ">u#s;W": [[]], "AIq3Wu": {}, "baz": [{}]}' else e'{"AwoIr\'\\"-": {"b": {"gBE7y7&`n7a{": {}}, "foo": "foo"}, "baz": {">bq)ecL": {"c": {}, "foo": true}, "foobar": [{"?gj`": {}}, {}], "sCTj>]U": [{"]#kE*tQK8": {}}]}}' end, NULL, 85.7738981053948509, ('242.91.52.50/23' - -7167524047846532832), '4.76.13.167/20', (e'{"3.]rKSLCCP`": [], "bar": {"D\\"0~y v": {"J-SbLKf": {"baz": {}}, "b": []}, "c": {}}, "c": null, "x6(i[": [1.7234824286955848], "zA@": {"(`1h6H(gZ": {}, "5pq": {"hDt": true}, "bar": {}, "yB_n6": "bar"}}' ->> 8692537976716570258), e'\\', (select (select distinct '1970-01-02 05:31:57.000674+00:00' from table0 as tab_4 limit 1) from table0 as tab_1 join table2 as tab_2 on ('61 years 417 days 15:46:51.848548' != (select '-49 years -1 mons -669 days -06:28:52.895501' from table5 as tab_3 limit 1)) limit 1), e'[true, {"a": null, "baz": []}, [{"A.[U9iIX": "|_d\'M58-"}, 0.64802256353997, [{}]], null, 0.4303066304677673]', NULL, date_recv('{"bar": [], "c": false, "dl?%": {"foo": [{}]}}'), case when true then '134.171.206.242/23' else netmask('88.199.162.153/22') end, NULL), ((true <= NULL), cast(coalesce(B'11000101100101101100001011001000011010011001101110110101001000100001011011100010001110', B'10000111110100010001110000001110100011001010001101101101110111111101001110110111') as bit varying), ('1970-01-09 02:38:23.000199+00:00' IS NOT DISTINCT FROM '1970-01-03 19:03:21.000295+00:00'), '1971-02-04', ('' = ''), -0.8193437303188197, oidin('56 years 5 mons 13 days 19:03:06.130625'), e'[{"!oB||rDK": {}, "Xn*,I6$mgRWj": [{"Jgc\\\\vx|=": null}], "skkY6": [[]]}, true, {"ZlMi}6\\\\p9": "foo", "c": 0.04525799539128339}, []]', NULL, cluster_logical_timestamp(), ('120.184.138.102/14' + 3179453217232625953), (NULL + case when NULL then 'd127:3fad:b3c1:3b62:31af:79ff:883f:6867/3' else inet_client_addr() end), e'\x1dL', e'Ebh\x18+$~', '1970-01-07 11:23:39.000809+00:00', e'{"U\\"=lyi;fUeoY": true, "b": [{"c": "baz"}, []], "bar": {"c": true}, "foobar": null, "rqu_2#n*~s": {"*Zw)m:F>8|F2": false, "D{K0": {}, "bar": true}}', 232236316, '1984-11-26', '90.147.178.114/20', true);
```

```
panic: could not find overload for binary expression minus [recovered]
	panic: could not find overload for binary expression minus [recovered]
	panic: panic while executing 1 statements: VALUES (_, _, _, (_ + current_date()), _, CASE WHEN CASE WHEN _ THEN _ ELSE _ END THEN _ ELSE (CASE WHEN _ THEN _ ELSE _ END - CASE WHEN _ THEN _ ELSE ln(_) END) END, _, CASE WHEN _ THEN _ ELSE _ END, _, _, (_ - _), _, (_->>_), _, (SELECT (SELECT DISTINCT _ FROM _ AS _ LIMIT _) FROM _ AS _ JOIN _ AS _ ON (_ != (SELECT _ FROM _ AS _ LIMIT _)) LIMIT _), _, _, date_recv(_), CASE WHEN _ THEN _ ELSE netmask(_) END, _), (__more1__); caused by could not find overload for binary expression minus

goroutine 352 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc0018ae000, 0x383dc80, 0xc001501700, 0x2ba9220, 0xc000486cb0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:721 +0x36d
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc0018ae000, 0x383dc80, 0xc001501700)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:427 +0x61
panic(0x2ba9220, 0xc000486cb0)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).Build.func1(0xc0014c0988)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:141 +0x9a
panic(0x2ba9220, 0xc000486cb0)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.BinaryAllowsNullArgs(0xc00041005e, 0x385b460, 0x587ade0, 0x385b460, 0x587ade0, 0xc0014bfcb0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/typing.go:80 +0x132
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*CustomFuncs).AllowNullArgs(0xc0018aeba0, 0x587005e, 0x385fa20, 0xc00041f900, 0x385fa20, 0xc00041f900, 0x2e04a80)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/custom_funcs.go:1084 +0x93
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).ConstructMinus(0xc0018aeb90, 0x385fa20, 0xc00041f900, 0x385fa20, 0xc00041f900, 0xc0014bfd01, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:11613 +0x4d6
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).constructBinary(0xc001296960, 0x4, 0x385fa20, 0xc00041f900, 0x385fa20, 0xc00041f900, 0x385b040, 0x587ade0, 0xc00041f900, 0xc001438c60)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/scalar.go:629 +0x248
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildScalar(0xc001296960, 0x3858460, 0xc001436400, 0xc0013ea000, 0x0, 0x0, 0x0, 0x385e820, 0xc001438c60)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/scalar.go:195 +0xc7a
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildScalar(0xc001296960, 0x38584c0, 0xc0017748c0, 0xc0013ea000, 0x0, 0x0, 0x0, 0x385eb20, 0x587b200)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/scalar.go:222 +0x2e0d
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildValuesClause(0xc001296960, 0xc000402840, 0x0, 0x0, 0x0, 0xc0013ea000, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/values.go:68 +0x361
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildSelect(0xc001296960, 0xc001232580, 0x0, 0x0, 0x0, 0xc0013ea000, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:588 +0x6cc
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildStmt(0xc001296960, 0x3843780, 0xc001232580, 0xc0013ea000, 0x4)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:203 +0x10e
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).Build(0xc001296960, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:148 +0xf6
github.com/cockroachdb/cockroach/pkg/sql.(*optPlanningCtx).buildExecMemo(0xc000a4b500, 0x383dd40, 0xc001438b40, 0xc001232580, 0x383dd40, 0xc001438b40, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:363 +0x18b
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makeOptimizerPlan(0xc0018ae458, 0x383dd40, 0xc001438b40, 0x30b7099, 0x19, 0x1d6d15163ab7, 0x36015ce6)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:145 +0xc8
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).makeExecPlan(0xc0018ae000, 0x383dd40, 0xc001438b40, 0xc0018ae458, 0x6, 0x69bc6f)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:930 +0x11b
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc0018ae000, 0x383dd40, 0xc001438b40, 0xc0018ae458, 0x7fa826c6d910, 0xc0006791e0, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:825 +0x148
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc0018ae000, 0x383dd40, 0xc001438b40, 0x3843780, 0xc001232580, 0xc0018d5000, 0x9b7, 0x0, 0x0, 0x0, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:438 +0xe49
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc0018ae000, 0x383dd40, 0xc001438b40, 0x3843780, 0xc001232580, 0xc0018d5000, 0x9b7, 0x0, 0x0, 0x0, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:93 +0x36f
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc0018ae000, 0x383dc80, 0xc001501700, 0xc00077a9f8, 0x5400, 0x15000, 0xc00077aa90, 0xc0015197e0, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1182 +0x21e5
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc000aefd40, 0x383dc80, 0xc001501700, 0xc0018ae000, 0x5400, 0x15000, 0xc00077aa90, 0xc0015197e0, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:429 +0xce
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func4(0xc000aefd40, 0xc0018a8000, 0xc0003af734, 0xc0015197e0, 0x383dc80, 0xc001501700, 0xc0018ae000, 0x5400, 0x15000, 0xc00077aa90, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:331 +0xcf
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:314 +0x1040
```

--- a/pkg/sql/opt/norm/custom_funcs.go
+++ b/pkg/sql/opt/norm/custom_funcs.go
@@ -1139,15 +1139,28 @@ func (c *CustomFuncs) IsConstValueEqual(const1, const2 opt.ScalarExpr) bool {
 	}
 }
 
+// ensureTyped makes sure that any NULL passing through gets tagged with an
+// appropriate type.
+func (c *CustomFuncs) ensureTyped(d opt.ScalarExpr, typ types.T) opt.ScalarExpr {
+	if d.DataType() == types.Unknown {
+		return c.f.ConstructNull(typ)
+	}
+	return d
+}
+
 // SimplifyWhens removes known unreachable WHEN cases and constructs a new CASE
 // statement. Any known true condition is converted to the ELSE. If only the
 // ELSE remains, its expression is returned. condition must be a ConstValue.
 func (c *CustomFuncs) SimplifyWhens(
 	condition opt.ScalarExpr, whens memo.ScalarListExpr, orElse opt.ScalarExpr,
 ) opt.ScalarExpr {
+	var typ types.T
 	newWhens := make(memo.ScalarListExpr, 0, len(whens))
 	for _, item := range whens {
 		when := item.(*memo.WhenExpr)
+		if typ == nil || typ == types.Unknown {
+			typ = when.Typ
+		}
 		if opt.IsConstValueOp(when.Condition) {
 			if !c.IsConstValueEqual(condition, when.Condition) {
 				// Ignore known unmatching conditions.
@@ -1168,7 +1181,11 @@ func (c *CustomFuncs) SimplifyWhens(
 	// The ELSE value.
 	if len(newWhens) == 0 {
 		// ELSE is the only clause (there are no WHENs), remove the CASE.
-		return orElse
+		// NULLs in this position will not be typed, so we tag them with
+		// a type we observed earlier.
+		// typ will never be nil here because the definition of
+		// SimplifyCaseWhenConstValue ensures that whens is nonempty.
+		return c.ensureTyped(orElse, typ)
 	}
 
 	return c.f.ConstructCase(condition, newWhens, orElse)

--- a/pkg/sql/opt/norm/testdata/rules/scalar
+++ b/pkg/sql/opt/norm/testdata/rules/scalar
@@ -597,6 +597,42 @@ values
  ├── fd: ()-->(1)
  └── ('one',) [type=tuple{string}]
 
+opt expect=SimplifyCaseWhenConstValue
+SELECT CASE 1 WHEN 2 THEN 'one' END
+----
+values
+ ├── columns: case:1(string)
+ ├── cardinality: [1 - 1]
+ ├── key: ()
+ ├── fd: ()-->(1)
+ └── (NULL,) [type=tuple{string}]
+
+opt expect=SimplifyCaseWhenConstValue
+SELECT CASE 1 WHEN 2 THEN 'one' ELSE NULL END
+----
+values
+ ├── columns: case:1(string)
+ ├── cardinality: [1 - 1]
+ ├── key: ()
+ ├── fd: ()-->(1)
+ └── (NULL,) [type=tuple{string}]
+
+# Regression test for #34930.
+norm
+SELECT
+    CASE
+    WHEN true THEN NULL
+    ELSE -0.41697856420581636
+    END
+    - CASE WHEN NULL THEN 1.4034371360919229 ELSE ln(NULL) END
+----
+values
+ ├── columns: "?column?":1(decimal)
+ ├── cardinality: [1 - 1]
+ ├── key: ()
+ ├── fd: ()-->(1)
+ └── (NULL,) [type=tuple{decimal}]
+
 # Verify that a true condition does not remove non-constant expressions
 # proceeding it.
 opt expect=SimplifyCaseWhenConstValue

